File: packages/core/src/event.ts
import type {
  DispatchConfig,
} from "./config";
import type {
  DispatchEvent,
  DispatchExceptionValue,
  DispatchFrame,
  DispatchRequest,
  DispatchUser,
  Level,
  Tags,
} from "./types";
import { parseStack } from "./stacktrace";
import { epochSeconds, uuid32 } from "./util";

export const MAX_CAUSES = 5;
export const VALUE_MAX_LENGTH = 2000;

export interface BuildEventInput {
  config: DispatchConfig;
  handled: boolean;
  platform?: string;
  level?: Level;
  user?: DispatchUser | null;
  tags?: Tags;
  request?: DispatchRequest;
  transaction?: string | null;
  serverName?: string | null;
  /** Injectable for runtime-specific parsing (e.g. the Node V8 parser). Defaults to core's. */
  parseStack?: (stack?: string | null) => DispatchFrame[];
  /** Injectable for deterministic tests. */
  now?: () => number;
  uuid?: () => string;
}

interface ErrorLike {
  name?: unknown;
  message?: unknown;
  stack?: unknown;
  cause?: unknown;
}

function asError(value: unknown): ErrorLike | null {
  return typeof value === "object" && value !== null ? (value as ErrorLike) : null;
}

function exceptionValue(
  error: unknown,
  handled: boolean,
  parse: (stack?: string | null) => DispatchFrame[],
): DispatchExceptionValue {
  const err = asError(error);
  const type = (err && typeof err.name === "string" && err.name) || "Error";
  const rawMessage =
    err && typeof err.message === "string" ? err.message : String(error ?? "");
  const stack = err && typeof err.stack === "string" ? err.stack : undefined;
  return {
    type,
    value: rawMessage.slice(0, VALUE_MAX_LENGTH),
    mechanism: { type: "generic", handled },
    stacktrace: { frames: parse(stack) },
  };
}

// Walk the `cause` chain up to MAX_CAUSES and emit Sentry-ordered values: OLDEST CAUSE FIRST,
// the raised error LAST. Mirrors EventBuilder#exception_values.
function exceptionValues(
  error: unknown,
  handled: boolean,
  parse: (stack?: string | null) => DispatchFrame[],
): DispatchExceptionValue[] {
  const chain: unknown[] = [];
  let current: unknown = error;
  for (let depth = 0; depth < MAX_CAUSES && current != null; depth++) {
    chain.push(current);
    current = asError(current)?.cause;
  }
  chain.reverse();
  return chain.map((e) => exceptionValue(e, handled, parse));
}

// Build the Sentry-shaped event from an error. The pure assembly lives here; runtime-specific
// concerns (source context, request extraction) are layered on by injecting parseStack and
// passing request/user/transaction.
export function buildEvent(error: unknown, input: BuildEventInput): DispatchEvent {
  const parse = input.parseStack ?? parseStack;
  const now = input.now ?? epochSeconds;
  const id = input.uuid ?? uuid32;
  const c = input.config;

  const derivedTags: Tags = {};
  if (input.transaction) derivedTags["transaction"] = input.transaction;
  const tags: Tags = { ...derivedTags, ...c.tags, ...(input.tags ?? {}) };

  const event: DispatchEvent = {
    event_id: id(),
    timestamp: now(),
    platform: input.platform ?? "javascript",
    level: input.level ?? "error",
    environment: c.environment,
    exception: { values: exceptionValues(error, input.handled, parse) },
  };

  // Optional fields are set only when present, so they're omitted (not null) on the wire.
  if (c.release != null) event.release = c.release;
  if (input.serverName != null) event.server_name = input.serverName;
  if (input.transaction != null) event.transaction = input.transaction;
  if (input.request) event.request = input.request;
  const user = input.user ?? c.user;
  if (user != null) event.user = user;
  if (Object.keys(tags).length > 0) event.tags = tags;
  event.sdk = c.sdk;

  return event;
}

File: python/dispatch-sdk/src/dispatch_sdk/event.py
"""Build the Sentry-shaped event from an exception. Mirrors the gem's EventBuilder."""

from __future__ import annotations

import time
import uuid as _uuid
from typing import Any, Callable, Dict, List, Optional

from .config import Config
from .stacktrace import frames_from_traceback
from .types import DispatchEvent, ExceptionValue
from .util import compact

MAX_CAUSES = 5
VALUE_MAX_LENGTH = 2000


def _exception_type_name(exc: BaseException) -> str:
    t = type(exc)
    module = getattr(t, "__module__", "") or ""
    if module in ("builtins", "__main__", ""):
        return t.__qualname__
    return f"{module}.{t.__qualname__}"


def _cause_of(exc: BaseException) -> Optional[BaseException]:
    """Follow the exception chain the way Python prints it: explicit `raise ... from` wins,
    otherwise the implicit context (unless suppressed). Uses typed attribute access so the
    chain stays BaseException|None rather than Any."""
    if exc.__cause__ is not None:
        return exc.__cause__
    if not exc.__suppress_context__:
        return exc.__context__
    return None


def _exception_value(exc: BaseException, handled: bool, project_root: str) -> ExceptionValue:
    tb = getattr(exc, "__traceback__", None)
    return {
        "type": _exception_type_name(exc),
        "value": str(exc)[:VALUE_MAX_LENGTH],
        "mechanism": {"type": "generic", "handled": handled},
        "stacktrace": {"frames": frames_from_traceback(tb, project_root)},
    }


def _exception_values(
    exc: BaseException, handled: bool, project_root: str
) -> List[ExceptionValue]:
    chain: List[BaseException] = []
    current: Optional[BaseException] = exc
    depth = 0
    while current is not None and depth < MAX_CAUSES:
        chain.append(current)
        current = _cause_of(current)
        depth += 1
    chain.reverse()  # oldest cause first, raised exception last
    return [_exception_value(e, handled, project_root) for e in chain]


def build_event(
    exc: BaseException,
    *,
    config: Config,
    handled: bool,
    platform: str = "python",
    level: str = "error",
    user: Optional[Dict[str, Any]] = None,
    tags: Optional[Dict[str, Any]] = None,
    request: Optional[Dict[str, Any]] = None,
    transaction: Optional[str] = None,
    server_name: Optional[str] = None,
    now: Callable[[], float] = time.time,
    uuid: Callable[[], str] = lambda: _uuid.uuid4().hex,
) -> DispatchEvent:
    derived: Dict[str, Any] = {}
    if transaction:
        derived["transaction"] = transaction
    merged_tags: Dict[str, Any] = {**derived, **config.tags, **(tags or {})}

    event: DispatchEvent = {
        "event_id": uuid(),
        "timestamp": now(),
        "platform": platform,
        "level": level,
        "environment": config.environment,
        "exception": {"values": _exception_values(exc, handled, config.project_root)},
    }
    event.update(
        compact(
            {
                "release": config.release,
                "server_name": server_name,
                "transaction": transaction,
                "request": request,
                "user": user if user is not None else config.user,
                "tags": merged_tags if merged_tags else None,
                "sdk": {"name": config.sdk_name, "version": config.sdk_version},
            }
        )
    )
    return event
